In [246]:
import pandas as pd
import numpy as np
from acquire_prepare import get_data_sql, prepare, parse_log_entry, summarize

In [48]:
df = get_data_sql()

In [49]:
# drops nulls
df = df.drop('deleted_at', axis=1).dropna()

In [50]:
# creates series of list from path columns
split_path = df.path.str.split('/')

In [51]:
# drops 'content' from each list
for ls in split_path:
    if ls[0] == 'content':
        ls.remove('content')

# drops 'further-reading' from each list
for ls in split_path:
    if len(ls) > 1 and ls[1] == 'further-reading':
        ls.remove('further-reading')

In [52]:
# parses the remaining lists in dplit path and puts them in the originfal dataframe as new columns
df[['topic', 'lesson']] = split_path.apply(parse_log_entry)

In [54]:
# converts the date and time to correct dtype and overwrites old date column
df['date'] = pd.to_datetime((df['date'] +' '+ df['time']))

# drops the unneeded time col
df.drop(['time', 'id'], axis=1, inplace=True)

# change cohort id to an integer
df['cohort_id'] = df.cohort_id.astype(int)

In [56]:
# how can I filter the df so that I only get the dates and times for when they were active
# convert the date and times for strart and end date to datetime
df['start_date'] = pd.to_datetime(df['start_date'])
df['end_date'] = pd.to_datetime(df['end_date'])

In [55]:
df.head()

date                                 path  user_id  \
0 2018-01-26 09:55:03                                    /        1   
1 2018-01-26 09:56:02                              java-ii        1   
2 2018-01-26 09:56:05  java-ii/object-oriented-programming        1   
3 2018-01-26 09:56:06   slides/object_oriented_programming        1   
4 2018-01-26 09:56:24            javascript-i/conditionals        2   

   cohort_id            ip     name     slack  start_date    end_date  \
0          8  97.105.19.61  Hampton  #hampton  2015-09-22  2016-02-06   
1          8  97.105.19.61  Hampton  #hampton  2015-09-22  2016-02-06   
2          8  97.105.19.61  Hampton  #hampton  2015-09-22  2016-02-06   
3          8  97.105.19.61  Hampton  #hampton  2015-09-22  2016-02-06   
4         22  97.105.19.61    Teddy    #teddy  2018-01-08  2018-05-17   

            created_at           updated_at  program_id         topic  \
0  2016-06-14 19:52:26  2016-06-14 19:52:26           1                 
1  2016-06-14 19:52:26  2016-06-14 19:52:26           1       java-ii   
2  2016-06-14 19:52:26  2016-06-14 19:52:26           1       java-ii   
3  2016-06-14 19:52:26  2016-06-14 19:52:26           1        slides   
4  2018-01-08 13:59:10  2018-01-08 13:59:10           2  javascript-i   

                        lesson  
0                               
1                               
2  object-oriented-programming  
3  object_oriented_programming  
4                 conditionals

### 1. Which lesson appears to attract the most traffic consistently across cohorts (per program)?
> #### Plan:
> * create separate df for each program (make it digestible)
> * check the amount of cohorts in each
> * groupby by cohorts
> * count how many time a lesson has been clicked on here
> * for each cohort grouping sort descending
> * get the head for each

In [5]:
# checking data
df.head()

date                                 path  user_id  \
0 2018-01-26 09:55:03                                    /        1   
1 2018-01-26 09:56:02                              java-ii        1   
2 2018-01-26 09:56:05  java-ii/object-oriented-programming        1   
3 2018-01-26 09:56:06   slides/object_oriented_programming        1   
4 2018-01-26 09:56:24            javascript-i/conditionals        2   

   cohort_id            ip     name  program_id         topic  \
0          8  97.105.19.61  Hampton           1                 
1          8  97.105.19.61  Hampton           1       java-ii   
2          8  97.105.19.61  Hampton           1       java-ii   
3          8  97.105.19.61  Hampton           1        slides   
4         22  97.105.19.61    Teddy           2  javascript-i   

                        lesson  
0                               
1                               
2  object-oriented-programming  
3  object_oriented_programming  
4                 conditionals

In [280]:
# creating different df for each program id
# what are the differnt program_ids
df.program_id.value_counts()

# separating
program_one = df[df['program_id'] == 1]
program_two = df[df['program_id'] == 2]
program_three = df[df['program_id'] == 3]
program_four = df[df['program_id'] == 4]

In [281]:
# program one
pd.DataFrame(program_one.groupby('cohort_id')['lesson'].value_counts())[program_one.groupby('cohort_id')['lesson'].value_counts()>900]\
.rename({'lesson':'lesson_count'}, axis=1)\
.sort_values('lesson_count', ascending=False)

lesson_count
cohort_id lesson              
14                        2780
1                         1831
17                         958

In [65]:
# program one
# pd.DataFrame(program_one.groupby(['cohort_id', 'topic'])['topic'].value_counts())[program_one.groupby('cohort_id')['topic'].value_counts()>900]\
# .rename({'topic':'topic_count'}, axis=1)\
# .sort_values('topic_count', ascending=False)

In [282]:

# program one
pd.DataFrame(program_one.groupby(['cohort_id', 'topic'])['cohort_id', 'topic'].value_counts())\
[program_one.groupby('cohort_id')['topic'].value_counts()>900]\
.sort_values(['cohort_id', 0], ascending=False)\
.groupby('cohort_id').head(1)

/var/folders/fj/7fbysvvn7956npcj9d4mxfxw0000gn/T/ipykernel_1889/563413257.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  pd.DataFrame(program_one.groupby(['cohort_id', 'topic'])['cohort_id', 'topic'].value_counts())\
/var/folders/fj/7fbysvvn7956npcj9d4mxfxw0000gn/T/ipykernel_1889/563413257.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  pd.DataFrame(program_one.groupby(['cohort_id', 'topic'])['cohort_id', 'topic'].value_counts())\


0
cohort_id topic             
17        javascript-i   953
14        spring        1104
1         javascript-i  1333

In [283]:
# program two
pd.DataFrame(program_two.groupby(['cohort_id', 'topic'])['cohort_id', 'topic'].value_counts())\
[program_two.groupby('cohort_id')['topic'].value_counts()>1000]\
.sort_values(['cohort_id', 0], ascending=False)\
.groupby('cohort_id').head(1)

/var/folders/fj/7fbysvvn7956npcj9d4mxfxw0000gn/T/ipykernel_1889/532094218.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  pd.DataFrame(program_two.groupby(['cohort_id', 'topic'])['cohort_id', 'topic'].value_counts())\
/var/folders/fj/7fbysvvn7956npcj9d4mxfxw0000gn/T/ipykernel_1889/532094218.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  pd.DataFrame(program_two.groupby(['cohort_id', 'topic'])['cohort_id', 'topic'].value_counts())\


0
cohort_id topic             
139       javascript-i  1331
138       javascript-i  2902
135       javascript-i  5849
134       javascript-i  3644
132       javascript-i  4049
62        javascript-i  5364
61        javascript-i  2904
58        javascript-i  5147
57        javascript-i  4852
56        javascript-i  4436
53        javascript-i  5167
52        html-css      4028
51        javascript-i  5038
33        javascript-i  6994
32        javascript-i  5462
31        javascript-i  4077
29        javascript-i  5860
28                      6340
27        javascript-i  3249
26        javascript-i  4289
25        javascript-i  4634
24        javascript-i  5001
23        javascript-i  4273
22        mysql         4439
21        spring        1772

In [57]:
# program three
pd.DataFrame(program_three.groupby(['cohort_id', 'topic'])['cohort_id', 'topic'].value_counts())\
[program_three.groupby('cohort_id')['topic'].value_counts()>1000]\
.sort_values(['cohort_id', 0], ascending=False)\
.groupby('cohort_id').head(2)

/var/folders/fj/7fbysvvn7956npcj9d4mxfxw0000gn/T/ipykernel_1391/3559874815.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  pd.DataFrame(program_three.groupby(['cohort_id', 'topic'])['cohort_id', 'topic'].value_counts())\
/var/folders/fj/7fbysvvn7956npcj9d4mxfxw0000gn/T/ipykernel_1391/3559874815.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  pd.DataFrame(program_three.groupby(['cohort_id', 'topic'])['cohort_id', 'topic'].value_counts())\


0
cohort_id topic               
137       fundamentals    3451
          sql             2277
133       classification  2715
          fundamentals    2469
59        classification  5030
                          2980
55        3-sql           2499
          4-python        2330
34        1-fundamentals  3124
          6-regression    2573

In [64]:
# program four
pd.DataFrame(program_four.groupby(['cohort_id', 'topic'])['cohort_id', 'topic'].value_counts())\
.sort_values(['cohort_id', 0], ascending=False)\
.groupby('cohort_id').head(2)

/var/folders/fj/7fbysvvn7956npcj9d4mxfxw0000gn/T/ipykernel_1391/722415719.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  pd.DataFrame(program_four.groupby(['cohort_id', 'topic'])['cohort_id', 'topic'].value_counts())\


0
cohort_id topic      
9         html-css  4
                    1

#### Takeaways
* Program 1: javascript-i
* Program 2: javasvript-i
* Program 3: fundamentals, closely followed by classification
* Program 4: html-css

### 2. Is there a cohort that referred to a lesson significantly more than other cohorts seemed to gloss over?
> #### Plan:
> * What is significance in this case? 
> * For each lesson, we can calculate an overall average of for each cohort. For each overall average, we can identify if there is

### 3. Are there students who, when active, hardly access the curriculum? If so, what information do you have about these students?
> #### Plan:
> * First, we need to find out what 'when active' means. Each cohort must be filtered for only their active dates. 
> * After this, I can answer this question by grouping by the user ids. With each user id, I can take a count of how many times they occur in the dataframe. After taking the count, I can get summary statistics about each, using the iqr to identify when 

In [45]:
# checking col names
df.head()

date                                 path  user_id  \
0 2018-01-26 09:55:03                                    /        1   
1 2018-01-26 09:56:02                              java-ii        1   
2 2018-01-26 09:56:05  java-ii/object-oriented-programming        1   
3 2018-01-26 09:56:06   slides/object_oriented_programming        1   
4 2018-01-26 09:56:24            javascript-i/conditionals        2   

   cohort_id            ip     name     slack  start_date    end_date  \
0          8  97.105.19.61  Hampton  #hampton  2015-09-22  2016-02-06   
1          8  97.105.19.61  Hampton  #hampton  2015-09-22  2016-02-06   
2          8  97.105.19.61  Hampton  #hampton  2015-09-22  2016-02-06   
3          8  97.105.19.61  Hampton  #hampton  2015-09-22  2016-02-06   
4         22  97.105.19.61    Teddy    #teddy  2018-01-08  2018-05-17   

            created_at           updated_at  program_id topic lesson  
0  2016-06-14 19:52:26  2016-06-14 19:52:26           1     /         
1  2016-06-14 19:52:26  2016-06-14 19:52:26           1     j      a  
2  2016-06-14 19:52:26  2016-06-14 19:52:26           1     j      a  
3  2016-06-14 19:52:26  2016-06-14 19:52:26           1     s      l  
4  2018-01-08 13:59:10  2018-01-08 13:59:10           2     j      a

In [50]:
# start date and end date values
(df.start_date.value_counts() == df.start_date.value_counts()).value_counts()

True    44
Name: start_date, dtype: int64

In [53]:
# there are 47 cohorts and 44 start and end dates
len(df.cohort_id.value_counts())

47

In [62]:
# next, i will focus on one cohort
df.groupby('cohort_id')['user_id'].nunique().idxmax()

28

something is wrong with cohort 28
* they have too many user and all the start and end dats are the same

In [57]:
# set index to date
play = df.copy()
play.set_index('date', inplace=True)

In [58]:
# nothing showing up?
mask = (play.cohort_id == 2)
play[mask][(play[mask].index == play[mask]['end_date'])]

Empty DataFrame
Columns: [path, user_id, cohort_id, ip, name, slack, start_date, end_date, created_at, updated_at, program_id, topic, lesson]
Index: []

In [59]:
# creating new df to hold activity during active times in the program
active_df = pd.DataFrame(columns=play.columns)

In [60]:
# checking that its empty
active_df

Empty DataFrame
Columns: [path, user_id, cohort_id, ip, name, slack, start_date, end_date, created_at, updated_at, program_id, topic, lesson]
Index: []

In [61]:
# creating mask and setup for getting the df
mask = (play.cohort_id == 31)
addition = play[mask][(play[mask].index >= play[mask]['start_date']) & (play[mask].index <= play[mask]['end_date'])]

In [64]:
# mergind the data
active_df = pd.concat([active_df, addition])

In [66]:
# active_df

In [74]:
# new clean df
active_df = pd.DataFrame(columns=play.columns)

# For each cohort in cohort_id
for i in play.cohort_id.unique():
    
    # mask it with the cohort
    mask = (play.cohort_id == i)

    # get all values for the active user during this time
    filtered_addition = play[mask][(play[mask].index >= play[mask]['start_date']) & (play[mask].index <= play[mask]['end_date'])]
    
    # concatinating the data onto the blank df for each
    active_df = pd.concat([active_df, filtered_addition])

In [76]:
# this is a df with all activity from users while they were activiely in the program
# now we can analyze this info
active_df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 647197 entries, 2018-01-26 09:56:24 to 2021-04-21 13:14:10
Data columns (total 13 columns):
 #   Column      Non-Null Count   Dtype         
---  ------      --------------   -----         
 0   path        647197 non-null  object        
 1   user_id     647197 non-null  object        
 2   cohort_id   647197 non-null  object        
 3   ip          647197 non-null  object        
 4   name        647197 non-null  object        
 5   slack       647197 non-null  object        
 6   start_date  647197 non-null  datetime64[ns]
 7   end_date    647197 non-null  datetime64[ns]
 8   created_at  647197 non-null  object        
 9   updated_at  647197 non-null  object        
 10  program_id  647197 non-null  object        
 11  topic       647197 non-null  object        
 12  lesson      647197 non-null  object        
dtypes: datetime64[ns](2), object(11)
memory usage: 69.1+ MB


In [77]:
def value_counts_and_frequencies(s: pd.Series, dropna=True) -> pd.DataFrame:
    return pd.merge(
        s.value_counts(dropna=False).rename('count'),
        s.value_counts(dropna=False, normalize=True).rename('proba'),
        left_index=True,
        right_index=True,
    )

In [211]:
# getting the values and frequencies 
user_df = value_counts_and_frequencies(active_df['user_id'])

In [212]:
# check
user_df.head()

count     proba
344   4763  0.007359
570   4075  0.006296
685   3223  0.004980
671   3193  0.004934
555   3086  0.004768

In [213]:
# getting users that barely accessed the data
user_df['count'].nsmallest(n = 50, keep='all')

879     1
619     1
918     1
940     1
832     3
278     4
539     5
956     6
812     7
388     8
679    11
572    12
941    12
697    13
889    16
968    17
487    17
963    18
340    18
961    20
49     25
64     25
772    26
24     26
41     28
976    29
972    34
973    35
529    36
959    41
967    41
981    42
371    42
971    44
964    45
741    47
960    48
56     52
543    53
46     54
978    55
965    55
962    58
57     62
858    63
45     65
36     65
975    67
250    67
969    71
Name: count, dtype: int64

In [113]:
# getting the count of the bottom 10 percent
under_ten_perc = user_df['count'].quantile(.10)
# getting the user_id's with less than 10 percent of clicks
users_under_ten_ls = list(user_df[user_df['count'] < under_ten_perc].index)

In [123]:
# some cohorts have many user_ids with low amounts, I'm guesing that some are people who did not stay in the program while others may be something of concer
active_df[active_df['user_id'].isin(users_under_ten_ls)]\
.groupby('cohort_id')['user_id'].nunique()

cohort_id
21     12
22      2
23      1
24      4
27      1
29      2
31      3
33      1
34      1
52      3
53      1
57      2
58      1
59      2
61      2
62      3
132     1
133     1
135     2
138     6
139    22
Name: user_id, dtype: int64

In [144]:
# get the last click per user and compare it to the start and end date
final_use_10_perc = active_df[active_df['user_id'].isin(users_under_ten_ls)]\
.sort_index()\
.groupby('user_id').tail(1)

In [152]:
final_use_10_perc['end_date_diff'] = final_use_10_perc.index - final_use_10_perc['end_date']

In [153]:
final_use_10_perc['start_date_diff'] = final_use_10_perc.index - final_use_10_perc['start_date']

In [154]:
# majority
summarize(final_use_10_perc)

                    SUMMARY REPORT


Dataframe head: 


path user_id cohort_id  \
2018-02-05 09:08:24  spring/fundamentals/form-model-binding      49        21   
2018-02-05 10:40:44     spring/extra-features/json-response      45        21   
2018-02-05 15:37:36                            mysql/tables      55        21   

                                  ip     name     slack start_date   end_date  \
2018-02-05 09:08:24     97.105.19.61  Sequoia  #sequoia 2017-09-27 2018-02-15   
2018-02-05 10:40:44  104.193.225.250  Sequoia  #sequoia 2017-09-27 2018-02-15   
2018-02-05 15:37:36     97.105.19.61  Sequoia  #sequoia 2017-09-27 2018-02-15   

                              created_at           updated_at program_id  \
2018-02-05 09:08:24  2017-09-27 20:22:41  2017-09-27 20:22:41          2   
2018-02-05 10:40:44  2017-09-27 20:22:41  2017-09-27 20:22:41          2   
2018-02-05 15:37:36  2017-09-27 20:22:41  2017-09-27 20:22:41          2   

                      topic          lesson      end_date_diff  \
2018-02-05 09:08:24  spring    fundamentals -10 days +09:08:24   
2018-02-05 10:40:44  spring  extra-features -10 days +10:40:44   
2018-02-05 15:37:36   mysql          tables -10 days +15:37:36   

                      start_date_diff  
2018-02-05 09:08:24 131 days 09:08:24  
2018-02-05 10:40:44 131 days 10:40:44  
2018-02-05 15:37:36 131 days 15:37:36



Dataframe info: 
<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 73 entries, 2018-02-05 09:08:24 to 2021-04-21 12:32:01
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype          
---  ------           --------------  -----          
 0   path             73 non-null     object         
 1   user_id          73 non-null     object         
 2   cohort_id        73 non-null     object         
 3   ip               73 non-null     object         
 4   name             73 non-null     object         
 5   slack            73 non-null     object         
 6   start_date       73 non-null     datetime64[ns] 
 7   end_date         73 non-null     datetime64[ns] 
 8   created_at       73 non-null     object         
 9   updated_at       73 non-null     object         
 10  program_id       73 non-null     object         
 11  topic            73 non-null     object         
 12  lesson           73 non-null     object         
 13  end_date_diff    73 non-n

Empty DataFrame
Columns: []
Index: []



Dataframe Description: 


count                          mean  \
end_date_diff      73 -108 days +12:28:23.082191780   
start_date_diff    73    53 days 01:17:41.986301370   

                                       std                 min  \
end_date_diff   71 days 09:52:23.678743152 -188 days +14:54:20   
start_date_diff 58 days 15:31:53.575997394     0 days 14:10:17   

                                25%                 50%                75%  \
end_date_diff   -164 days +21:08:32 -150 days +14:10:17 -13 days +14:29:11   
start_date_diff     8 days 11:42:09     9 days 10:42:27  122 days 14:29:11   

                              max  
end_date_diff   -1 days +12:42:51  
start_date_diff 166 days 01:46:07

DataFrame value counts: 


path
/                                                      7
javascript-i/introduction/operators                    6
html-css/css-ii/bootstrap-grid-system                  5
javascript-i/conditionals                              5
html-css/elements                                      3
javascript-i/introduction/primitive-types              3
spring/fundamentals/security/authentication            3
appendix                                               2
javascript-i/testing-user-functions                    2
html-css/css-i/box-model                               2
spring                                                 2
appendix/capstone-workbook                             2
mysql/tables                                           2
javascript-i/loops                                     2
javascript-i/javascript-with-html                      2
html-css                                               2
javascript-i/introduction/working-with-data-typ...     1
spring/fundamentals/form-model-binding                 1
java-i/strings                                         1
fundamentals/git                                       1
toc                                                    1
javascript-i/objects                                   1
search/search_index.json                               1
appendix/git/cli                                       1
appendix/cli/6-moving-files                            1
1-fundamentals/modern-data-scientist.jpg               1
4-python/8.4.1-pandas-overview                         1
spring/fundamentals/views                              1
spring/extra-features/json-response                    1
javascript-i/mapbox-api                                1
html-css/css-i/positioning                             1
html-css/css-ii/media-queries                          1
appendix/git/intellij-intro                            1
java-ii/collections                                    1
java-i/console-io                                      1
spring/fundamentals/repositories                       1
jquery/events/mouse-events                             1
jquery/introduction                                    1
javascript-i/introduction/variables                    1

user_id
49         1
772        1
962        1
968        1
961        1
..       ...
432        1
431        1
371        1
388        1
969        1

[73 rows x 1 columns]

cohort_id
139         22
21          12
138          6
24           4
31           3
52           3
62           3
57           2
29           2
135          2
22           2
59           2
61           2
133          1
132          1
23           1
58           1
33           1
34           1
27           1
53           1

ip
97.105.19.58     11
97.105.19.61      6
136.50.89.186     2
71.221.46.112     2
69.239.143.192    1
70.121.220.245    1
70.121.129.79     1
136.50.53.26      1
162.200.114.251   1
50.27.36.194      1
67.11.98.144      1
72.183.145.159    1
99.186.6.129      1
72.193.254.235    1
96.8.168.216      1
107.204.52.180    1
173.174.134.234   1
67.11.222.229     1
98.156.147.200    1
70.117.2.210      1
75.31.192.132     1
72.177.209.77     1
97.117.115.55     1
68.203.215.174    1
73.85.64.161      1
72.26.31.59       1
70.123.219.174    1
172.58.97.72      1
173.175.19.30     1
104.193.225.250   1
107.77.217.110    1
173.173.100.228   1
24.160.133.222    1
141.150.59.235    1
67.11.154.74      1
24.155.140.118    1
173.174.159.45    1
104.53.215.21     1
76.214.156.188    1
107.77.217.9      1
66.69.108.117     1
64.148.251.171    1
70.125.148.96     1
136.50.50.187     1
12.106.208.194    1
136.50.70.27      1
24.28.146.155     1
76.221.87.98      1
75.58.161.186     1
76.187.101.165    1
24.243.12.201     1
24.243.49.105     1
99.162.243.68     1
69.154.52.98      1
130.45.49.89      1
107.77.169.64     1

name
Oberon       22
Sequoia      12
Neptune       6
Voyageurs     4
Andromeda     3
Europa        3
Jupiter       3
Ganymede      2
Zion          2
Marco         2
Teddy         2
Darden        2
Bash          2
Easley        1
Kalypso       1
Ulysses       1
Hyperion      1
Ceres         1
Bayes         1
Yosemite      1
Fortuna       1

slack
#oberon        22
#sequoia       12
#neptune        6
#voyageurs      4
#andromeda      3
#europa         3
#jupiter        3
#ganymede       2
#zion           2
#marco          2
#teddy          2
#darden         2
#bash           2
#easley         1
#kalypso        1
#ulysses        1
#hyperion       1
#ceres          1
#               1
#yosemite       1
#fortuna        1

start_date
(2017-09-25 16:58:04.799999999, 2018-02-03 07:1...          14
(2018-02-03 07:12:00, 2018-06-12 14:24:00]                   5
(2018-06-12 14:24:00, 2018-10-19 21:36:00]                   0
(2018-10-19 21:36:00, 2019-02-26 04:48:00]                   3
(2019-02-26 04:48:00, 2019-07-05 12:00:00]                   3
(2019-07-05 12:00:00, 2019-11-11 19:12:00]                   5
(2019-11-11 19:12:00, 2020-03-20 02:24:00]                   1
(2020-03-20 02:24:00, 2020-07-27 09:36:00]                   7
(2020-07-27 09:36:00, 2020-12-03 16:48:00]                   4
(2020-12-03 16:48:00, 2021-04-12]                           31

end_date
(2018-02-13 16:13:26.399999999, 2018-06-27 09:3...        14
(2018-06-27 09:36:00, 2018-11-06 19:12:00]                 5
(2018-11-06 19:12:00, 2019-03-19 04:48:00]                 0
(2019-03-19 04:48:00, 2019-07-29 14:24:00]                 3
(2019-07-29 14:24:00, 2019-12-09]                          3
(2019-12-09, 2020-04-19 09:36:00]                          5
(2020-04-19 09:36:00, 2020-08-29 19:12:00]                 3
(2020-08-29 19:12:00, 2021-01-09 04:48:00]                 1
(2021-01-09 04:48:00, 2021-05-21 14:24:00]                 8
(2021-05-21 14:24:00, 2021-10-01]                         31

created_at
2021-04-12 18:07:21          22
2017-09-27 20:22:41          12
2021-03-15 19:57:09           6
2018-05-25 22:25:57           4
2019-03-18 20:35:06           3
2019-11-04 18:27:07           3
2020-09-21 18:06:27           3
2020-03-23 17:52:16           2
2019-01-20 23:18:57           2
2021-01-20 21:31:11           2
2018-01-08 13:59:10           2
2020-07-13 18:32:19           2
2020-07-20 19:04:00           2
2020-12-07 15:20:18           1
2020-11-02 20:43:58           1
2018-03-05 14:22:11           1
2020-05-26 19:22:44           1
2019-07-15 16:57:21           1
2019-08-20 14:38:55           1
2018-11-05 15:26:37           1
2020-01-13 21:17:08           1

updated_at
2021-04-12 18:07:21          22
2017-09-27 20:22:41          12
2021-03-15 19:57:09           6
2018-05-25 22:25:57           4
2019-03-18 20:35:06           3
2019-11-04 18:27:07           3
2020-09-21 18:06:27           3
2020-03-23 17:52:16           2
2019-01-20 23:18:57           2
2021-01-20 21:31:11           2
2018-01-08 13:59:10           2
2020-07-13 18:32:19           2
2020-09-30 15:54:46           2
2020-12-07 15:20:18           1
2020-11-02 20:43:58           1
2018-03-05 14:22:11           1
2020-05-26 19:22:44           1
2019-07-15 16:57:21           1
2019-08-20 14:38:55           1
2018-11-05 15:26:37           1
2020-01-13 21:17:08           1

program_id
2          69
3           4

topic
javascript-i       24
html-css           14
spring              9
appendix            7
                    7
mysql               2
jquery              2
java-i              2
java-ii             1
4-python            1
1-fundamentals      1
fundamentals        1
toc                 1
search              1

lesson
introduction                   12
                                7
                                7
fundamentals                    6
css-ii                          6
conditionals                    5
git                             3
elements                        3
css-i                           3
loops                           2
capstone-workbook               2
javascript-with-html            2
tables                          2
testing-user-functions          2
console-io                      1
collections                     1
events                          1
extra-features                  1
mapbox-api                      1
cli                             1
8.4.1-pandas-overview           1
modern-data-scientist.jpg       1
objects                         1
strings                         1
search_index.json               1

end_date_diff
(-188 days +10:25:11.088999999, -169 days +07:2...              9
(-169 days +07:29:11.100000, -150 days +00:04:0...             27
(-150 days +00:04:02.200000, -132 days +16:38:5...              7
(-132 days +16:38:53.300000, -113 days +09:13:4...              3
(-113 days +09:13:44.400000, -94 days +01:48:35...              2
(-94 days +01:48:35.500000, -76 days +18:23:26....              2
(-76 days +18:23:26.600000, -57 days +10:58:17....              0
(-57 days +10:58:17.700000, -38 days +03:33:08....              0
(-38 days +03:33:08.800000, -20 days +20:07:59....              0
(-20 days +20:07:59.900000, -1 days +12:42:51]                 23

start_date_diff
(0 days 10:11:59.249999999, 17 days 03:19:52]               41
(17 days 03:19:52, 33 days 16:29:27]                         0
(33 days 16:29:27, 50 days 05:39:02]                         5
(50 days 05:39:02, 66 days 18:48:37]                         0
(66 days 18:48:37, 83 days 07:58:12]                         3
(83 days 07:58:12, 99 days 21:07:47]                         1
(99 days 21:07:47, 116 days 10:17:22]                        1
(116 days 10:17:22, 132 days 23:26:57]                       8
(132 days 23:26:57, 149 days 12:36:32]                      13
(149 days 12:36:32, 166 days 01:46:07]                       1

nulls in dataframe by column: 


num_rows_missing  percent_rows_missing
path                            0                   0.0
user_id                         0                   0.0
cohort_id                       0                   0.0
ip                              0                   0.0
name                            0                   0.0
slack                           0                   0.0
start_date                      0                   0.0
end_date                        0                   0.0
created_at                      0                   0.0
updated_at                      0                   0.0
program_id                      0                   0.0
topic                           0                   0.0
lesson                          0                   0.0
end_date_diff                   0                   0.0
start_date_diff                 0                   0.0

nulls in dataframe by row: 


num_cols_missing  percent_cols_missing
2018-02-05 09:08:24                 0                   0.0
2020-09-23 14:54:20                 0                   0.0
2021-04-20 16:28:32                 0                   0.0
2021-04-20 15:25:42                 0                   0.0
2021-04-20 12:04:53                 0                   0.0
...                               ...                   ...
2019-07-29 12:42:51                 0                   0.0
2019-07-22 11:13:31                 0                   0.0
2019-03-29 13:36:04                 0                   0.0
2019-03-19 12:19:23                 0                   0.0
2021-04-21 12:32:01                 0                   0.0

[73 rows x 2 columns]

#### Takeaways
* Majority of the last uses end at day 17 (41 of them)
    * This may be students who decided not to continue with the program
* Oberon (139) and Sequoia (21) have an abnormally high number of users (22, 12) that use it extremely little and discontinue use later on 
* Vast majority of the users are in web development programs, with 4 in a datascience cohort

In [87]:
def get_lower_and_upper_bounds(df, col, k=1.5):
    '''
    Takes in a df and a column names and return the lower fence and upper fence with a k of 1.5
    '''
    # first quartile
    q1 = df[col].quantile(.25)
    
    # second quartile
    q3 = df[col].quantile(.75)
    
    # calculate the iqr
    iqr = (q3 - q1)
    
    # get the upper fence
    upper_fence = q3 + (k * iqr)
    
    # get the lower fence
    lower_fence = q1 - (k * iqr)
    
    # exit and return the upper and lower fence
    return lower_fence, upper_fence

In [91]:
get_lower_and_upper_bounds(user_df, 'count')\
, get_lower_and_upper_bounds(user_df, 'proba')

((-874.5, 2501.5), (-0.0013512114549356692, 0.003865129164690195))

In [100]:
x = user_df['count']
# Calculate the z-score 
zscores = pd.Series((x - x.mean()) / x.std())

# Finds all of the observations two standard deviations or more.
x[zscores <= -1.2]

# Finds all of the observations three standard deviations or more
x[zscores.abs() >= 3]

344    4763
570    4075
685    3223
671    3193
555    3086
764    2955
423    2946
823    2946
760    2923
794    2852
Name: count, dtype: int64

In [101]:
user_df.describe()

count       proba
count   731.000000  731.000000
mean    885.358413    0.001368
std     646.913125    0.001000
min       1.000000    0.000002
25%     391.500000    0.000605
50%     795.000000    0.001228
75%    1235.500000    0.001909
max    4763.000000    0.007359

### 4. Is there any suspicious activity, such as users/machines/etc accessing the curriculum who shouldn’t be? Does it appear that any web-scraping is happening? Are there any suspicious IP addresses?
> #### Plan:
> * What defines suspicious activity?
> * What does webscraping look like in tabular data?
> * What consititutes suspicious ip addresses?

Yes, there is quite a bit of suspicious activity happening
There are also a lot of suspivious ip addresses that come up a lot

In [179]:
def acquire(file_name, column_names):
    return pd.read_csv(file_name, sep="\s", header=None, names=column_names, usecols=[0, 2, 3, 4, 5])

def prep(df, user):
    df = df[df.user_id == user]
    pages = df['path'].resample('d').count()
    return pages

def compute_pct_b(pages, span, weight, user):
    midband = pages.ewm(span=span).mean()
    stdev = pages.ewm(span=span).std()
    ub = midband + stdev*weight
    lb = midband - stdev*weight
    bb = pd.concat([ub, lb], axis=1)
    my_df = pd.concat([pages, midband, bb], axis=1)
    my_df.columns = ['pages', 'midband', 'ub', 'lb']
    my_df['pct_b'] = (my_df['pages'] - my_df['lb'])/(my_df['ub'] - my_df['lb'])
    my_df['user_id'] = user
    return my_df

def plt_bands(my_df, user):
    fig, ax = plt.subplots(figsize=(12,8))
    ax.plot(my_df.index, my_df.pages, label='Number of Pages, User: '+str(user))
    ax.plot(my_df.index, my_df.midband, label = 'EMA/midband')
    ax.plot(my_df.index, my_df.ub, label = 'Upper Band')
    ax.plot(my_df.index, my_df.lb, label = 'Lower Band')
    ax.legend(loc='best')
    ax.set_ylabel('Number of Pages')
    plt.show()

def find_anomalies(df, user, span, weight):
    pages = prep(df, user)
    my_df = compute_pct_b(pages, span, weight, user)
    # plt_bands(my_df, user)
    return my_df[my_df.pct_b>1]


In [216]:
user = 1
span = 30
weight = 6
user_anomaly_df = find_anomalies(df, user, span, weight)

anomalies = pd.DataFrame()
user_anomaly_df = find_anomalies(df, user, span, weight)
anomalies = pd.concat([anomalies, user_df], axis=0)

In [218]:
span = 30
weight = 3.5

anomalies = pd.DataFrame()
for u in list(df.user_id.unique()):
    user_anomaly_df = find_anomalies(df, u, span, weight)
    anomalies = pd.concat([anomalies, user_anomaly_df], axis=0)

In [219]:
page_anomaly_ld = list(anomalies.pages.value_counts().nsmallest(keep='all').index)

In [220]:
user_anomalies_ls = list(anomalies[anomalies.pages.isin(page_anomaly_ld)]['user_id'].values)

In [221]:
df[df['user_id'].isin(user_anomalies_ls)].groupby('user_id').count()

path  cohort_id    ip  name  slack  start_date  end_date  created_at  \
user_id                                                                         
33        151        151   151   151    151         151       151         151   
72        530        530   530   530    530         530       530         530   
101       224        224   224   224    224         224       224         224   
138      1264       1264  1264  1264   1264        1264      1264        1264   
153       760        760   760   760    760         760       760         760   
223      1316       1316  1316  1316   1316        1316      1316        1316   
248      5075       5075  5075  5075   5075        5075      5075        5075   
255       114        114   114   114    114         114       114         114   
264      2083       2083  2083  2083   2083        2083      2083        2083   
309       809        809   809   809    809         809       809         809   
315      4510       4510  4510  4510   4510        4510      4510        4510   
341      1581       1581  1581  1581   1581        1581      1581        1581   
344      5460       5460  5460  5460   5460        5460      5460        5460   
375      1633       1633  1633  1633   1633        1633      1633        1633   
423      3804       3804  3804  3804   3804        3804      3804        3804   
430      2811       2811  2811  2811   2811        2811      2811        2811   
434      1795       1795  1795  1795   1795        1795      1795        1795   
480      2303       2303  2303  2303   2303        2303      2303        2303   
482      1391       1391  1391  1391   1391        1391      1391        1391   
521      3052       3052  3052  3052   3052        3052      3052        3052   
526      1225       1225  1225  1225   1225        1225      1225        1225   
556      1446       1446  1446  1446   1446        1446      1446        1446   
560      1219       1219  1219  1219   1219        1219      1219        1219   
564      1680       1680  1680  1680   1680        1680      1680        1680   
567       732        732   732   732    732         732       732         732   
581      6434       6434  6434  6434   6434        6434      6434        6434   
609      1941       1941  1941  1941   1941        1941      1941        1941   
635      1476       1476  1476  1476   1476        1476      1476        1476   
642       894        894   894   894    894         894       894         894   
658      1654       1654  1654  1654   1654        1654      1654        1654   
664      1153       1153  1153  1153   1153        1153      1153        1153   
785       986        986   986   986    986         986       986         986   

         updated_at  program_id  topic  lesson  date  
user_id                                               
33              151         151    151     151   151  
72              530         530    530     530   530  
101             224         224    224     224   224  
138            1264        1264   1264    1264  1264  
153             760         760    760     760   760  
223            1316        1316   1316    1316  1316  
248            5075        5075   5075    5075  5075  
255             114         114    114     114   114  
264            2083        2083   2083    2083  2083  
309             809         809    809     809   809  
315            4510        4510   4510    4510  4510  
341            1581        1581   1581    1581  1581  
344            5460        5460   5460    5460  5460  
375            1633        1633   1633    1633  1633  
423            3804        3804   3804    3804  3804  
430            2811        2811   2811    2811  2811  
434            1795        1795   1795    1795  1795  
480            2303        2303   2303    2303  2303  
482            1391        1391   1391    1391  1391  
521            3052        3052   3052    3052  3052  
526            1225        1225   1225    12

In [236]:
df[df['user_id'].isin(user_anomalies_ls)].ip.value_counts(normalize=True)

97.105.19.58       0.322147
70.112.179.142     0.066937
173.173.115.201    0.042256
72.190.187.173     0.030371
99.88.62.179       0.029184
                     ...   
64.134.155.129     0.000016
174.128.180.24     0.000016
107.77.217.49      0.000016
24.173.208.242     0.000016
107.77.221.36      0.000016
Name: ip, Length: 318, dtype: float64

In [238]:
suspicious_ls = user_df[user_df.index.isin(user_anomalies_ls)].index

In [242]:
df[df['user_id'].isin(suspicious_ls)].ip.value_counts()

97.105.19.58      17583
70.112.179.142     4117
72.190.187.173     1868
99.88.62.179       1795
75.1.166.163       1769
                  ...  
99.203.213.177        1
107.77.220.123        1
12.5.63.210           1
107.77.221.36         1
174.207.25.208        1
Name: ip, Length: 278, dtype: int64

In [233]:
# anomalies.user_id user_df.index

### 5. At some point in 2019, the ability for students and alumni to access both curriculums (web dev to ds, ds to web dev) should have been shut off. Do you see any evidence of that happening? Did it happen before?
> #### Plan:
> * Identify which program is ds and which is wd, add column called program_type
> * Identify which paths are ds and which are wd, create a column called path_type
> * Identify when the change in ability happen
> * Check to see if there have been anytimes that program_type != path_type

In [264]:
df['in_webdev'] = np.where((df['program_id'] == 1), 1,\
         np.where((df['program_id'] == 2), 1,\
                  np.where((df['program_id'] == 4), 1, 0)))

array([1, 1, 1, ..., 1, 1, 1])

In [274]:
# yes, this is happening
df[['in_webdev', 'path']]

in_webdev  \
date                             
2018-01-26 09:55:03          1   
2018-01-26 09:56:02          1   
2018-01-26 09:56:05          1   
2018-01-26 09:56:06          1   
2018-01-26 09:56:24          1   
...                        ...   
2021-04-21 16:41:51          1   
2021-04-21 16:42:02          1   
2021-04-21 16:42:09          1   
2021-04-21 16:44:37          1   
2021-04-21 16:44:39          1   

                                                                 path  
date                                                                   
2018-01-26 09:55:03                                                 /  
2018-01-26 09:56:02                                           java-ii  
2018-01-26 09:56:05               java-ii/object-oriented-programming  
2018-01-26 09:56:06                slides/object_oriented_programming  
2018-01-26 09:56:24                         javascript-i/conditionals  
...                                                               ...  
2021-04-21 16:41:51                              jquery/personal-site  
2021-04-21 16:42:02                                 jquery/mapbox-api  
2021-04-21 16:42:09                           jquery/ajax/weather-map  
2021-04-21 16:44:37  anomaly-detection/discrete-probabilistic-methods  
2021-04-21 16:44:39                                 jquery/mapbox-api  

[847329 rows x 2 columns]

### 6. What topics are grads continuing to reference after graduation and into their jobs (for each program)?
> #### Plan:
> * Create a df with only post-graduation curriculum accesses
> * Get value_counts from those times

In [301]:
# new clean df
inactive_df = pd.DataFrame(columns=play.columns)

# For each cohort in cohort_id
for i in df.cohort_id.unique():
    
    # mask it with the cohort
    mask = (play.cohort_id == i)

    # get all values for the active user during this time
    filtered_addition = play[mask][play[mask].index > play[mask]['end_date']]
    
    # concatinating the data onto the blank df for each
    inactive_df = pd.concat([inactive_df, filtered_addition])

In [302]:
inactive_df.head()

path user_id cohort_id  \
2018-01-26 09:55:03                                    /       1         8   
2018-01-26 09:56:02                              java-ii       1         8   
2018-01-26 09:56:05  java-ii/object-oriented-programming       1         8   
2018-01-26 09:56:06   slides/object_oriented_programming       1         8   
2018-01-26 10:40:15               javascript-i/functions       1         8   

                               ip     name     slack start_date   end_date  \
2018-01-26 09:55:03  97.105.19.61  Hampton  #hampton 2015-09-22 2016-02-06   
2018-01-26 09:56:02  97.105.19.61  Hampton  #hampton 2015-09-22 2016-02-06   
2018-01-26 09:56:05  97.105.19.61  Hampton  #hampton 2015-09-22 2016-02-06   
2018-01-26 09:56:06  97.105.19.61  Hampton  #hampton 2015-09-22 2016-02-06   
2018-01-26 10:40:15  97.105.19.61  Hampton  #hampton 2015-09-22 2016-02-06   

                              created_at           updated_at program_id  \
2018-01-26 09:55:03  2016-06-14 19:52:26  2016-06-14 19:52:26          1   
2018-01-26 09:56:02  2016-06-14 19:52:26  2016-06-14 19:52:26          1   
2018-01-26 09:56:05  2016-06-14 19:52:26  2016-06-14 19:52:26          1   
2018-01-26 09:56:06  2016-06-14 19:52:26  2016-06-14 19:52:26          1   
2018-01-26 10:40:15  2016-06-14 19:52:26  2016-06-14 19:52:26          1   

                            topic                       lesson  
2018-01-26 09:55:03                                             
2018-01-26 09:56:02       java-ii                               
2018-01-26 09:56:05       java-ii  object-oriented-programming  
2018-01-26 09:56:06        slides  object_oriented_programming  
2018-01-26 10:40:15  javascript-i                    functions

In [303]:
# separating
program_one = inactive_df[inactive_df['program_id'] == 1]
program_two = inactive_df[inactive_df['program_id'] == 2]
program_three = inactive_df[inactive_df['program_id'] == 3]
program_four = inactive_df[inactive_df['program_id'] == 4]

In [305]:
# checking
program_one.head()

path user_id cohort_id  \
2018-01-26 09:55:03                                    /       1         8   
2018-01-26 09:56:02                              java-ii       1         8   
2018-01-26 09:56:05  java-ii/object-oriented-programming       1         8   
2018-01-26 09:56:06   slides/object_oriented_programming       1         8   
2018-01-26 10:40:15               javascript-i/functions       1         8   

                               ip     name     slack start_date   end_date  \
2018-01-26 09:55:03  97.105.19.61  Hampton  #hampton 2015-09-22 2016-02-06   
2018-01-26 09:56:02  97.105.19.61  Hampton  #hampton 2015-09-22 2016-02-06   
2018-01-26 09:56:05  97.105.19.61  Hampton  #hampton 2015-09-22 2016-02-06   
2018-01-26 09:56:06  97.105.19.61  Hampton  #hampton 2015-09-22 2016-02-06   
2018-01-26 10:40:15  97.105.19.61  Hampton  #hampton 2015-09-22 2016-02-06   

                              created_at           updated_at program_id  \
2018-01-26 09:55:03  2016-06-14 19:52:26  2016-06-14 19:52:26          1   
2018-01-26 09:56:02  2016-06-14 19:52:26  2016-06-14 19:52:26          1   
2018-01-26 09:56:05  2016-06-14 19:52:26  2016-06-14 19:52:26          1   
2018-01-26 09:56:06  2016-06-14 19:52:26  2016-06-14 19:52:26          1   
2018-01-26 10:40:15  2016-06-14 19:52:26  2016-06-14 19:52:26          1   

                            topic                       lesson  
2018-01-26 09:55:03                                             
2018-01-26 09:56:02       java-ii                               
2018-01-26 09:56:05       java-ii  object-oriented-programming  
2018-01-26 09:56:06        slides  object_oriented_programming  
2018-01-26 10:40:15  javascript-i                    functions

In [324]:
# program 
pd.DataFrame(program_one.groupby(['cohort_id', 'topic'])['cohort_id', 'topic'].value_counts())\
[program_one.groupby('cohort_id')['topic'].value_counts()>100]\
.sort_values(['cohort_id', 0], ascending=False)\
.groupby('cohort_id').head(2)

/var/folders/fj/7fbysvvn7956npcj9d4mxfxw0000gn/T/ipykernel_1889/1388460011.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  pd.DataFrame(program_one.groupby(['cohort_id', 'topic'])['cohort_id', 'topic'].value_counts())\
/var/folders/fj/7fbysvvn7956npcj9d4mxfxw0000gn/T/ipykernel_1889/1388460011.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  pd.DataFrame(program_one.groupby(['cohort_id', 'topic'])['cohort_id', 'topic'].value_counts())\


0
cohort_id topic             
19        laravel        187
          mysql          182
17        javascript-i   953
          html-css       636
14        spring        1104
          javascript-i  1006
13        laravel        591
          php_iv         273
8         java-iii       212
                         210
7         html-css       165
          javascript-i   150
1         javascript-i  1333
          html-css      1063

In [307]:
program_two.head()

path user_id cohort_id              ip  \
2018-05-17 10:30:24                       /      18        22    97.105.19.61   
2018-05-17 10:30:27                   mysql      18        22    97.105.19.61   
2018-05-17 10:30:41  mysql/basic-statements      18        22    97.105.19.61   
2018-05-17 11:20:04                       /       6        22    97.105.19.61   
2018-05-17 13:57:07                  jquery       6        22  64.132.225.203   

                      name   slack start_date   end_date           created_at  \
2018-05-17 10:30:24  Teddy  #teddy 2018-01-08 2018-05-17  2018-01-08 13:59:10   
2018-05-17 10:30:27  Teddy  #teddy 2018-01-08 2018-05-17  2018-01-08 13:59:10   
2018-05-17 10:30:41  Teddy  #teddy 2018-01-08 2018-05-17  2018-01-08 13:59:10   
2018-05-17 11:20:04  Teddy  #teddy 2018-01-08 2018-05-17  2018-01-08 13:59:10   
2018-05-17 13:57:07  Teddy  #teddy 2018-01-08 2018-05-17  2018-01-08 13:59:10   

                              updated_at program_id   topic            lesson  
2018-05-17 10:30:24  2018-01-08 13:59:10          2                            
2018-05-17 10:30:27  2018-01-08 13:59:10          2   mysql                    
2018-05-17 10:30:41  2018-01-08 13:59:10          2   mysql  basic-statements  
2018-05-17 11:20:04  2018-01-08 13:59:10          2                            
2018-05-17 13:57:07  2018-01-08 13:59:10          2  jquery

In [308]:
# program two
pd.DataFrame(program_two.groupby(['cohort_id', 'topic'])['cohort_id', 'topic'].value_counts())\
[program_two.groupby('cohort_id')['topic'].value_counts()>100]\
.sort_values(['cohort_id', 0], ascending=False)\
.groupby('cohort_id').head(1)

/var/folders/fj/7fbysvvn7956npcj9d4mxfxw0000gn/T/ipykernel_1889/134903616.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  pd.DataFrame(program_two.groupby(['cohort_id', 'topic'])['cohort_id', 'topic'].value_counts())\
/var/folders/fj/7fbysvvn7956npcj9d4mxfxw0000gn/T/ipykernel_1889/134903616.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  pd.DataFrame(program_two.groupby(['cohort_id', 'topic'])['cohort_id', 'topic'].value_counts())\


0
cohort_id topic             
61        spring         332
58        spring         252
57        javascript-i   594
56        java-ii        702
53        spring         464
52        javascript-i   323
51        javascript-i  1407
33        javascript-i   809
32        javascript-i   555
31        spring         962
29        html-css       688
28                      6340
27        spring         172
26        html-css       748
25        spring         251
24        spring         558
23        spring         626
22        spring         879
21        spring         783
18        javascript-i   256
16        spring         175
15        php_i          115

In [325]:
# program 4
pd.DataFrame(program_three.groupby(['cohort_id', 'path'])['cohort_id', 'path'].value_counts())\
[program_three.groupby('cohort_id')['path'].value_counts()>100]\
.sort_values(['cohort_id', 0], ascending=False)\
.groupby('cohort_id').head(3)

/var/folders/fj/7fbysvvn7956npcj9d4mxfxw0000gn/T/ipykernel_1889/3560508571.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  pd.DataFrame(program_three.groupby(['cohort_id', 'path'])['cohort_id', 'path'].value_counts())\
/var/folders/fj/7fbysvvn7956npcj9d4mxfxw0000gn/T/ipykernel_1889/3560508571.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  pd.DataFrame(program_three.groupby(['cohort_id', 'path'])['cohort_id', 'path'].value_counts())\


0
cohort_id path                                         
59        /                                         386
          search/search_index.json                  149
          sql/mysql-overview                        123
55        /                                         569
          search/search_index.json                  157
          classification/overview                   136
34        /                                         491
          search/search_index.json                  187
          1-fundamentals/modern-data-scientist.jpg  116

In [316]:
# program three
pd.DataFrame(program_four.groupby(['cohort_id', 'path'])['cohort_id', 'path'].value_counts())\
.sort_values(['cohort_id', 0], ascending=False)\
.groupby('cohort_id').head(3)

/var/folders/fj/7fbysvvn7956npcj9d4mxfxw0000gn/T/ipykernel_1889/1965904642.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  pd.DataFrame(program_four.groupby(['cohort_id', 'path'])['cohort_id', 'path'].value_counts())\


0
cohort_id path                                          
9         content/html-css                             2
          /                                            1
          content/html-css/gitbook/images/favicon.ico  1

### 7. Which lessons are least accessed?
> #### Plan:
> * 

In [323]:
# program one
pd.DataFrame(program_one.groupby(['cohort_id', 'path'])['cohort_id', 'path'].value_counts())\
.sort_values(['cohort_id', 0], ascending=False)\
.groupby('cohort_id').tail(1)

/var/folders/fj/7fbysvvn7956npcj9d4mxfxw0000gn/T/ipykernel_1889/1301382914.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  pd.DataFrame(program_one.groupby(['cohort_id', 'path'])['cohort_id', 'path'].value_counts())\


0
cohort_id path                                    
19        spring/setup                           1
17        spring/extra-features/json-response    1
14        spring/fundamentals/integration-tests  1
13        spring/setup                           1
12        prework/cli/02-listing-files           1
11        search/search_index.json               1
8         uploads/58a217a705bde.jpg              1
7         spring/fundamentals/controllers        1
6         spring/fundamentals/controllers        1
5         /                                      1
4         prework/versioning/github              1
2         prework/fundamentals/variable          1
1         students/units/75/sub_units/268        1

In [322]:
# program two
pd.DataFrame(program_two.groupby(['cohort_id', 'path'])['cohort_id', 'path'].value_counts())\
.sort_values(['cohort_id', 0], ascending=False)\
.groupby('cohort_id').tail(1)

/var/folders/fj/7fbysvvn7956npcj9d4mxfxw0000gn/T/ipykernel_1889/2376498038.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  pd.DataFrame(program_two.groupby(['cohort_id', 'path'])['cohort_id', 'path'].value_counts())\


0
cohort_id path                                
62        web-design/ux/purpose              1
61        slides/syntax_types_and_variables  1
58        web-design/ux/purpose              1
57        web-design/ux/purpose              1
56        web-design/ux/purpose              1
53        web-design/ux/purpose              1
52        web-design/ui/visuals              1
51        syntax-types-and-variables         1
33        web-design/ux/purpose              1
32        web-design/ux/purpose              1
31        web-design/ux/layout               1
29        web-design/ux/purpose              1
28        working-with-time-series-data      1
27        web-design/ui/visuals              1
26        web-design/ui/color                1
25        web-design/ux/purpose              1
24        web-design/ux/purpose              1
23        web-design/ui/typography           1
22        wp-login                           1
21        web-design/ux/purpose              1
18        web-design/ux/purpose              1
16        slides/user_input                  1
15        toc                                1

In [327]:
# program three
pd.DataFrame(program_three.groupby(['cohort_id', 'path'])['cohort_id', 'path'].value_counts())\
.sort_values(['cohort_id', 0], ascending=False)\
.groupby('cohort_id').tail(5)

/var/folders/fj/7fbysvvn7956npcj9d4mxfxw0000gn/T/ipykernel_1889/336019201.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  pd.DataFrame(program_three.groupby(['cohort_id', 'path'])['cohort_id', 'path'].value_counts())\


0
cohort_id path                                    
59        git                                    1
          individual-project/individual-project  1
          python/type-annotations                1
          regression/explore                     1
          stats/sampling                         1
55        storytelling/present                   1
          storytelling/refine                    1
          storytelling/tableau                   1
          storytelling/understand                1
          timeseries/svr                         1
34        storytelling/refine                    1
          storytelling/tableau                   1
          tidy-data                              1
          timeseries/svr                         1
          viz.py                                 1

### 8. Anything else I should be aware of?
> #### Plan:
> * 